styleguide-csharp.md

```diff
@@ -0,0 +1,113 @@
+##### 2.2.1.1 Configuration
+
+<a id="s2.2.1.1.1-annotation-vs-fluent-api"></a>
+###### 2.2.1.1.1 Annotations VS Fluent API
+
+- Prefer to use Fluent API over annotations where we can
+- Prefer to have one source of truth rather than split rules across different locations
+- Follow the following convention:
+    * Place all the configuration for an entity type inside a separate class implementing [`IEntityTypeConfiguration<TEntity>`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.ientitytypeconfiguration-1)
+    * Make this class `internal` and place it in the same file as where the entity class itself is defined
+    * All configurations within the assembly can then be registered from `OnModelCreating`:
+
+        ```cs
+        modelBuilder.ApplyConfigurationsFromAssembly(typeof(BlogEntityTypeConfiguration).Assembly);
+        ```
+
+<a id="s2.2.1.2-stored-procedures-vs-linq-vs-sql"></a>
+##### 2.2.1.2 Stored Procedures VS Linq VS SQL
+
+- Our general guidance is we try to do everything with EF
+- If we need additional performance or other SQL features unsupported by EF, we can consider generating SQL or a new stored procedure
+
+<a id="s2.3-best-practices"></a>
+### 2.3 Best Practices 
+
+<a id="s2.3.2-ef"></a>
+#### 2.3.2 Entity Framework (EF)
+
+- By default, use the synchronous versions of `Add`, `AddRange` etc. as opposed to the asynchronous versions `AddAsync`, `AddRangeAsync` etc.
+    * The asynchronous versions appear to only be useful in specialist situations. The Microsoft documentation suggests that it is only useful when using the [`HiLoValueGenerator`](https://miro.com/app/board/o9J_ly21jhs=/?moveToWidget=3458764574909560501&cot=14)
+- Prefer to use `context.{EntityType}.Add` over `context.Add`
+
+
+<a id="s3-csharp-testing-rules"></a>
+## 2 C# Testing Rules 
+
+<a id="s3.1-style"></a>
+### 3.1 Style 
+
+<a id="s3.1.1-test-naming-convention"></a>
+#### 3.1.1 Test Naming Convention
+
+- Top-level class names will follow the naming convention `{ClassUnderTest}Tests`
+- Create a nested class on a per-public-method being tested basis `{MethodUnderTest}Tests`
+    * Do this from the outset, from he first method being tested. Do not wait until there are mutliple methods being tested to convert to this style
+    * Also use doubly-nested classes on a per-feature basis where appropriate `{FeatureUnderTest}Tests`
+- Generally follow a `Scenario_Result` pattern for test method names
+- Simple happy-path scenario could just be called `Success` for succinctness
+
+
+<a id="s3.2-test-setup"></a>
+### 3.2 Test Setup 
+
+<a id="s3.3-best-practices"></a>
+### 3.3 Best Practices 
+
+<a id="s3.3.1-synchronous-vs-asynchronous-test-methods"></a>
+#### 3.3.1 Synchronous VS Asynchronous Test Methods
+
+- If a test methods is using async method(s), then the method declaration should be marked as `async Task` or `async Task<T>`. See [here](https://miro.com/app/board/o9J_ly21jhs=/?moveToWidget=3458764574908695514&cot=14).
```

Comment:
I'd assume that we wouldn't want contributor guides that referred to our internal Miro boards, so I'd be tempted to remove this link?

Reply:
Good point! I'll remove this

Reply:
Just looking back over this.... This was never actually meant to be a Miro link. It was intended to be a Microsoft link. I must have copied the wrong link!

I don't know what the original link was now though :')